File: Rust/src/donp/element/segment.rs
/*
	Package 'element' provides definitions for various objects meeting the definition of a single
	entity used within communication protocols.
*/
/*
	Objects of type Segment describe a single entity present within a communication protocol message.
	Required fields: name, bits
	Optional fields: desc
*/

use log::debug;

pub struct Segment {
    name: String,
    desc: String,
    bits: u8,    
}

impl Default for Segment {
    fn default() -> Self {
        Segment {
            name: String::new(),
            desc: String::new(),
            bits: 0,            
        }
    }
}

impl Clone for Segment {
    fn clone(&self) -> Segment {
        Segment {
            name: self.name.clone(),
            desc: self.desc.clone(),
            bits: self.bits,
        }
    }
}

impl AsMut<Segment> for Segment {
    fn as_mut(&mut self) -> &mut Segment {
        self
    }
}

impl Segment {
    pub fn new() -> Segment {
        Segment::default()
    }

    pub fn from_map(&mut self, map: &serde_json::Value) {
        // Initialize segment fields based on map.
        self.name = map["name"].as_str().unwrap_or("").to_string();
        self.desc = map["desc"].as_str().unwrap_or("").to_string();
        self.bits = map["bits"].as_u64().unwrap_or(0) as u8;
        self.log();
    }

    pub fn log(&self) {
        // Log segment details for debugging purposes.
        debug!("Segment Name: {} ({}), Bits: {}", self.name, self.desc, self.bits);
    }

    pub fn get_name(&self) -> &String {
        &self.name
    }

    pub fn get_bits(&self) -> u8 {
        self.bits
    }
}

File: C++/pkg/element/segment.cpp
/*
	Package 'element' provides definitions for various objects meeting the definition of a single
	entity used within communication protocols.
*/
/*
	Objects of type Segment describe a single entity present within a communication protocol message.
	Required fields: name, bits
	Optional fields: desc
*/


#include "pkg/element/segment.hpp"

// Constructor
Segment::Segment() {
    name = "";
	desc = "";
	bits = 0;
	logger = nullptr;
}

Segment::Segment(std::string json_config, std::shared_ptr<Logger> log_ptr) : logger(log_ptr) {
	auto view = glz::get_view_json<"/name">(json_config);	
	if (view) {
		name = std::string(view->begin() + 1, view->end() - 1);		
	}
	view = glz::get_view_json<"/desc">(json_config);
	if (view) {
		desc = std::string(view->begin() + 1, view->end() - 1);
	}
	view = glz::get_view_json<"/bits">(json_config);
	if (view) {
		bits = std::stoul(std::string(view->data(), view->size()));
	}
}

// Destructor
Segment::~Segment() {
    // Cleanup code if needed
}

// Methods
std::string Segment::get_name(void) const {
	return name;
}

std::string Segment::get_desc(void) const {
	return desc;
}

uint32_t Segment::get_bits(void) const {
	return bits;
}

void Segment::log(void) const {
	logger->log(DEBUG, "Segment Name: " + name + ", (" + desc + "), Bits: " + std::to_string(bits));
}	